pkg/api/api.go

```diff
@@ -240,6 +245,109 @@ func (a *API) RegisterDistributor(d *distributor.Distributor, pushConfig distrib
+
+		ctx := r.Context()
+		logger := util_log.WithContext(ctx, util_log.Logger)
+		if a.sourceIPs != nil {
+			source := a.sourceIPs.Get(r)
+			if source != "" {
+				ctx = util.AddSourceIPsToOutgoingContext(ctx, source)
+				logger = util_log.WithSourceIPs(source, logger)
+			}
+		}
+
+		if err := d.StartBackfill(ctx, tenantID, blockID); err != nil {
+			resp, ok := httpgrpc.HTTPResponseFromError(err)
+			if !ok {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			} else {
+				http.Error(w, string(resp.Body), int(resp.Code))
+			}
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}), true, false, http.MethodPost)
+	// Endpoint to handle requests for uploading block files to a backfill.
+	a.RegisterRoute("/api/v1/backfill/{tenant:[0-9]+}/{block}/{path}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		tenantID, err := strconv.Atoi(vars["tenant"])
+		if err != nil {
+			http.Error(w, fmt.Sprintf("invalid tenant ID: %q", vars["tenant"]), http.StatusBadRequest)
+			return
+		}
+		blockID := vars["block"]
+		pth, err := url.PathUnescape(vars["path"])
+		if err != nil {
+			http.Error(w, fmt.Sprintf("invalid file path: %q", vars["path"]), http.StatusBadRequest)
+			return
+		}
+
+		ctx := r.Context()
+		logger := util_log.WithContext(ctx, util_log.Logger)
+		if a.sourceIPs != nil {
+			source := a.sourceIPs.Get(r)
+			if source != "" {
+				ctx = util.AddSourceIPsToOutgoingContext(ctx, source)
+				logger = util_log.WithSourceIPs(source, logger)
+			}
+		}
+
+		if err := d.AddBackfillFile(ctx, tenantID, blockID, pth, r); err != nil {
+			resp, ok := httpgrpc.HTTPResponseFromError(err)
+			if !ok {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			} else {
+				http.Error(w, string(resp.Body), int(resp.Code))
+			}
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}), true, false, http.MethodPost)
+	// Endpoint to handle requests for finishing of block backfilling.
+	a.RegisterRoute("/api/v1/backfill/{tenant:[0-9]+}/{block}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		tenantID, err := strconv.Atoi(vars["tenant"])
+		if err != nil {
+			http.Error(w, fmt.Sprintf("invalid tenant ID: %q", vars["tenant"]), http.StatusBadRequest)
+			return
+		}
+		blockID := vars["block"]
+
+		ctx := r.Context()
+		logger := util_log.WithContext(ctx, util_log.Logger)
+		if a.sourceIPs != nil {
+			source := a.sourceIPs.Get(r)
+			if source != "" {
+				ctx = util.AddSourceIPsToOutgoingContext(ctx, source)
+				logger = util_log.WithSourceIPs(source, logger)
+			}
+		}
+
+		if err := d.FinishBackfill(ctx, tenantID, blockID); err != nil {
+			resp, ok := httpgrpc.HTTPResponseFromError(err)
+			if !ok {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			} else {
+				http.Error(w, string(resp.Body), int(resp.Code))
+			}
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+	}), true, false, http.MethodDelete)
```

Comment:
I don't think DELETE is a good method for signalling the end.

Reply:
This PR made me realize that what we're doing is very similar to S3 multipart upload ([learn more here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html)). For this reason, I would follow its naming, changing "Multipart" into "Block".

I propose to change the backfill API as follows:

- `CreateBlockUpload`:   `POST /api/v1/upload/block/{blockID}`
- `UploadBlockFile`:     `POST /api/v1/upload/block/{blockID}/{path}`
- `CompleteBlockUpload`: `POST /api/v1/upload/block/{blockID}?uploadComplete=true`

Notes:
- Removed the tenant ID from path
- I replaced `backfill/` with `upload/` because from the API perspective what we're really doing is just uploading a block (from the user perspective we call it "backfill" and it's fine, but in practice from the API perspective we just upload blocks)
- I suggest to add `block/` to the path, because in the future we may support other forms of uploads/backfilling

Reply:
Looks sensible to me.